File: 685-RedundantConnectionIi/685-RedundantConnectionIi.java
// Last updated: 8/3/2025, 9:54:13 AM
public class Solution {
    public int[] findRedundantDirectedConnection(int[][] edges) {
        int n = edges.length;
        int[] parent = new int[n + 1];
        int[] cand1 = null, cand2 = null;

        
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            if (parent[v] == 0) {
                parent[v] = u;
            } else {
               
                cand1 = new int[]{parent[v], v}; 
                cand2 = new int[]{u, v};          
                edge[1] = 0; 
            }
        }

      
        int[] root = new int[n + 1];
        for (int i = 1; i <= n; i++) root[i] = i;

        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            if (v == 0) continue; 

            int pu = find(root, u);
            if (pu == v) {
                
                return cand1 == null ? edge : cand1;
            }
            root[v] = pu;
        }

        return cand2; 
    }

    private int find(int[] root, int u) {
        if (root[u] != u)
            root[u] = find(root, root[u]);
        return root[u];
    }
}

File: 685-RedundantConnectionIi/685-RedundantConnectionIi.ex
# Last updated: 8/3/2025, 9:54:10 AM
defmodule Solution do
  def find_redundant_directed_connection(edges) do
    n = length(edges)
    parent = Map.new(1..n, fn i -> {i, i} end)

    {cand1, cand2, updated_edges, final_parent} =
      Enum.reduce(edges, {nil, nil, [], parent}, fn [u, v] = edge, {c1, c2, acc, pmap} ->
        if Map.get(pmap, v) != v do
          # Node has two parents
          {[Map.get(pmap, v), v], edge, acc ++ [[u, 0]], pmap}
        else
          {c1, c2, acc ++ [edge], Map.put(pmap, v, u)}
        end
      end)

    roots = Enum.reduce(1..n, %{}, fn i, acc -> Map.put(acc, i, i) end)

    {cycle_edge, _} =
      Enum.reduce_while(updated_edges, {nil, roots}, fn [u, v], {found, rmap} ->
        if v == 0 do
          {:cont, {found, rmap}}
        else
          {pu, rmap1} = find(rmap, u)
          {pv, rmap2} = find(rmap1, v)

          if pu == pv do
            {:halt, {[u, v], rmap2}}
          else
            {:cont, {found, Map.put(rmap2, v, pu)}}
          end
        end
      end)

    cond do
      cand1 == nil -> cycle_edge
      cycle_edge != nil -> cand1
      true -> cand2
    end
  end

  defp find(map, u) do
    if map[u] != u do
      {res, new_map} = find(map, map[u])
      {res, Map.put(new_map, u, res)}
    else
      {u, map}
    end
  end
end
